src/main/java/com/arpnetworking/tsdcore/sinks/HttpSinkActor.java

```diff
@@ -257,14 +269,37 @@ private void dispatchPending() {
     }
 
     private void fireNextRequest() {
-        final Request request = _pendingRequests.poll();
+        final RequestEntry requestEntry = _pendingRequests.poll();
+        final Metrics metrics = _metricsFactory.create();
+        metrics.setTimer(_inQueueLatencyName, System.currentTimeMillis() - requestEntry.getEnterTime(), TimeUnit.MILLISECONDS);
+
+        final Request request = requestEntry.getRequest();
         _inflightRequestsCount++;
 
         final CompletableFuture<Response> promise = new CompletableFuture<>();
+        metrics.startTimer(_requestLatencyName);
         _client.executeRequest(request, new ResponseAsyncCompletionHandler(promise));
         final CompletionStage<Object> responsePromise = promise
-                .<Object>thenApply(response -> new PostComplete(request, response))
-                .exceptionally(PostFailure::new);
+                .handle((result, err) -> {
+                        metrics.stopTimer(_requestLatencyName);
+                        final int responseStatusCode = result.getStatusCode();
```

Comment:
I think result may be null if err is not null; therefore this will NPE on an exception.

Reply:
The determination of success is:

If err == null then success = true
Else if status code in set then success = true
Else success = false (or set as default)

So this is the branch you have below. Why not just use that branch to determine and output success? You will need to refactor the multiple return points of course.